tuf/repository_lib.py

```diff
@@ -1549,6 +1566,222 @@ def _get_hashes_and_length_if_needed(use_length, use_hashes, full_file_path,
 
 
 
+class Node(object):
+  """
+  Merkle tree node that keeps track of the node digest and the parent node.
+  """
+  parent = None
+  digest = None
+
+  def __init__(self):
+    return
+
+  def is_leaf(self):
+    return False
+
+
+
+
+class InternalNode(Node):
+  """
+  An internal Merkle tree node that keeps track of a left and a right
+  child. Upon creation, this node takes in a left and right Node
+  and computes the digest of (left + right). In addition, the constructor
+  sets the parent node of left and right to this node to allow for
+  traversal of the tree.
+  """
+  left = None
+  right = None
+
+  def __init__(self, left, right):
+    super(InternalNode, self).__init__()
+    self.left = left
+    self.right = right
```

Comment:
Could we flatten the tree by allowing more than two children per internal node? And would it make sense to have a min or max key field included with each child to know how to search the tree?

Reply:
For a key field, I'm thinking something like "the timestamp of the target + content hash" which is unique while sorting nicely, older more static data to the earlier leaves and newer frequently changing data to the later leaves. That would hopefully have clients able to cache more of the data from previous queries.

Reply:
> Could we flatten the tree by allowing more than two children per internal node? And would it make sense to have a min or max key field included with each child to know how to search the tree?

I think we need not have an RTT per level of the tree.  This may be an implementation detail in the prototype we could clean up  The repository could bundle the full path to the leaf node and send it all in a single reply.  

Using a structure like you mention (which is like a B-Tree instead of a binary tree), will use more bandwidth, but require fewer signatures (on tree creation) and fewer blocks read from disk.  Our sense is that the bandwidth was the more important concern, but feel free to correct us!

Reply:
From the context of container registries, the round trips are an issue, and the non-leaves (index/manifests) tend to be small relative to the leaf nodes (layers/blobs). My thoughts are very much along the lines of a Merkle + B-tree since that maps so closely with container registries, but I'm missing the TUF context to know how that would result in significantly larger download sizes.